core/src/main/java/com/linecorp/armeria/internal/server/CorsHeaderUtil.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2024 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License
+ */
+
+package com.linecorp.armeria.internal.server;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Strings;
+
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.ResponseHeadersBuilder;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.cors.CorsConfig;
+import com.linecorp.armeria.server.cors.CorsPolicy;
+
+/**
+ * utility class related to CORS headers.
+ */
+public final class CorsHeaderUtil {
+
+    private static final Logger logger = LoggerFactory.getLogger(CorsHeaderUtil.class);
+
+    public static final String ANY_ORIGIN = "*";
+    public static final String NULL_ORIGIN = "null";
+
+    private CorsHeaderUtil() {
+    }
+
+    /**
+     * Emit CORS headers if origin was found.
+     *
+     * @param req     the HTTP request with the CORS info
+     * @param headers the headers to modify
+     */
+    public static void setCorsResponseHeaders(ServiceRequestContext ctx, HttpRequest req,
+                                              ResponseHeadersBuilder headers, CorsConfig config) {
+        final CorsPolicy policy = setCorsOrigin(ctx, req, headers, config, logger);
+        if (policy != null) {
+            setCorsAllowCredentials(headers, policy);
+            setCorsAllowHeaders(req.headers(), headers, policy);
+            setCorsExposeHeaders(headers, policy);
+        }
+    }
+
+    public static void setCorsAllowCredentials(ResponseHeadersBuilder headers, CorsPolicy policy) {
+        // The string "*" cannot be used for a resource that supports credentials.
+        // https://www.w3.org/TR/cors/#resource-requests
+        if (policy.isCredentialsAllowed() &&
+            !ANY_ORIGIN.equals(headers.get(HttpHeaderNames.ACCESS_CONTROL_ALLOW_ORIGIN))) {
+            headers.set(HttpHeaderNames.ACCESS_CONTROL_ALLOW_CREDENTIALS, "true");
+        }
+    }
+
+    public static void setCorsExposeHeaders(ResponseHeadersBuilder headers, CorsPolicy corsPolicy) {
```

Comment:
It seems we can make this method private.
```suggestion
    private static void setCorsExposeHeaders(ResponseHeadersBuilder headers, CorsPolicy corsPolicy) {
```